Lib/test/test_sqlite3/test_factory.py

```diff
@@ -251,6 +251,25 @@ def test_sqlite_row_keys(self):
         row = self.con.execute("select 1 as a, 2 as b").fetchone()
         self.assertEqual(row.keys(), ['a', 'b'])
 
+        # Test exception raised on an empty sql query result
+        with self.assertRaises(AttributeError):
+            row = self.con.execute("select 1 as a where a == 'THISDOESNOTEXIST'").fetchone()
+            row.keys()
```

Comment:
Can you narrow this down? Where is the AttributeError raised? Consider also using `assertRaisesRegex`.

Reply:
What is the type of `row`?

Reply:
So `.fetchone()` is returning None as per the comment below,

```
    # Returns either a row (as created by the row_factory) or None, but
    # putting None in the return annotation causes annoying false positives.
    def fetchone(self) -> Any: ...
```

and the type of `keys` has to be a list of strings,

```
class Row:
    def __init__(self, __cursor: Cursor, __data: tuple[Any, ...]) -> None: ...
    def keys(self) -> list[str]: ...
```

and here is the C code for the keys method,

```
static PyObject *
pysqlite_row_keys_impl(pysqlite_Row *self)
/*[clinic end generated code: output=efe3dfb3af6edc07 input=7549a122827c5563]*/
{
    PyObject* list;
    Py_ssize_t nitems, i;

    list = PyList_New(0);
    if (!list) {
        return NULL;
    }
    nitems = PyTuple_Size(self->description);

    for (i = 0; i < nitems; i++) {
        if (PyList_Append(list, PyTuple_GET_ITEM(PyTuple_GET_ITEM(self->description, i), 0)) != 0) {
            Py_DECREF(list);
            return NULL;
        }
    }

    return list;
}
```

however I'm not familiar with the codebase enough to go between the C code and the Python code to find where the error is being raised :(

Reply:
Since the query fails, no rows are returned, so `fetchone` returns `None`, so what you are really testing here is that `None` does not have a `keys` attribute :) I think we can do without this test.

Reply:
> What is the type of `row`?

None :) As already mentioned, we don't need this test.

Reply:
If `fetchone()` returns not a `Row`, this code has no any relation to testing `Row.keys()`.

Reply:
See https://github.com/python/cpython/pull/108628#discussion_r1308702602, Serhiy.